components/director/internal/systemfetcher/service.go

```diff
@@ -0,0 +1,195 @@
+		systemsAPIClient:   sac,
+		appTemplateService: appTemplateService,
+		workers:            make(chan struct{}, fetcherParallellism),
+	}
+}
+
+type TenantSystems struct {
+	tenant  *model.BusinessTenantMapping
+	systems []System
+}
+
+func (s *SystemFetcher) SyncSystems(ctx context.Context) error {
+	tenants, err := s.listTenants(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to list tenants")
+	}
+
+	systemsQueue := make(chan TenantSystems, 100)
+	wgDB := sync.WaitGroup{}
+	wgDB.Add(1)
+	go func() {
+		defer func() {
+			wgDB.Done()
+		}()
+		for tenantSystems := range systemsQueue {
+			err = s.saveSystemsForTenant(ctx, tenantSystems.tenant, tenantSystems.systems)
+			if err != nil {
+				log.C(ctx).Error(errors.Wrap(err, fmt.Sprintf("failed to save systems for tenant %s", tenantSystems.tenant.ExternalTenant)))
+				continue
+			}
+
+			log.C(ctx).Info(fmt.Sprintf("Successfully synced systems for tenant %s", tenantSystems.tenant.ExternalTenant))
+		}
+	}()
+
+	wg := sync.WaitGroup{}
+	for number, t := range tenants {
+		wg.Add(1)
+		s.workers <- struct{}{}
+		go func(t *model.BusinessTenantMapping, number int) {
+			defer func() {
+				wg.Done()
+				<-s.workers
+			}()
+			systems, err := s.systemsAPIClient.FetchSystemsForTenant(ctx, t.ExternalTenant)
+			if err != nil {
+				log.C(ctx).Error(errors.Wrap(err, fmt.Sprintf("failed to fetch systems for tenant %s", t.ExternalTenant)))
+				return
+			}
+			if len(systems) > 0 {
+				systemsQueue <- TenantSystems{
+					tenant:  t,
+					systems: systems,
+				}
+			}
+		}(t, number)
+	}
+
+	wg.Wait()
+	close(systemsQueue)
+	wgDB.Wait()
+
+	return nil
+}
+
+func (s *SystemFetcher) listTenants(ctx context.Context) ([]*model.BusinessTenantMapping, error) {
+	tx, err := s.transaction.Begin()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to begin transaction")
+	}
+	defer s.transaction.RollbackUnlessCommitted(ctx, tx)
+
+	ctx = persistence.SaveToContext(ctx, tx)
+
+	tenants, err := s.tenantService.List(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to retrieve tenants")
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to commit while retrieving tenants")
+	}
+
+	return tenants, nil
+}
+
+func (s *SystemFetcher) saveSystemsForTenant(ctx context.Context, tenantMapping *model.BusinessTenantMapping, systems []System) error {
+	var appInputs []model.ApplicationRegisterInput
+	var templateTypes []string
+
+	tx, err := s.transaction.Begin()
+	if err != nil {
+		return errors.Wrap(err, "failed to begin transaction")
+	}
+	defer s.transaction.RollbackUnlessCommitted(ctx, tx)
+
+	ctx = tenant.SaveToContext(ctx, tenantMapping.ID, tenantMapping.ExternalTenant)
+	ctx = persistence.SaveToContext(ctx, tx)
+
+	for _, system := range systems {
+		appInput, templateType := s.convertSystemToAppRegisterInput(system)
+		template, err := s.appTemplateService.GetByName(ctx, templateType)
+		if err != nil && !apperrors.IsNotFoundError(err) {
```

Comment:
Perhaps an error will tell us that we have a non-existing template mapping. As a second thought this can be done at the startup and convert once the template mapping to IDs and there will no need to get them one by one here...